README
signature
=========

Ever wanted to have witty and pithy sigs attached to your e-mail and
news postings, but got tired having the same old one attached every
time?

That's where signature comes in. signature is a free, open-source
producer of dynamic signatures for livening up your e-mail and news
postings. It will allow you to sign your messages with a different sig
every time.

The idea for the program came from an old Perl script in the Camel
Book (http://www.oreilly.com/catalog/pperl2/) that demonstrated the
use of named pipes.

This C implementation was written as an exercise in programming. I'm an
experienced UNIX sysadmin, not a professional programmer, although back
in the dark ages I used to program for pleasure quite a lot in BASIC
(back when they still used line numbers), Pascal, Modula-2 and Z80
assembler.

In recent years, I've used Perl more or less exclusively, and - since
it's such a flexible language - I just never got further than the
basics of C. I decided to change all that recently and have been
doing a lot of reading and sifting through C code.

signature is pretty much the first fruit to come of these labours.

Since this program is the work of a novice in C, please excuse any
golden rules of programming that I've inadvertently broken in the
code. Do feel free to inform me of them, however, and I'll endeavour
to find a way to program more elegantly.

signature has been successfully compiled and tested on Linux
2.2.x. Early versions were also known to work on Solaris 2.6. Maybe
the current one still does, but I no longer have access to such a box
to test it.

I'd like signature to be portable if at all possible, so if there's a
platform you'd like it to work on, let me know what's required (or
better still, send a diff).

The program was developed on an x86 Red Hat Linux 6.2 system.

Sources consulted during the development of this software include:

'Beginning Linux Programming' by Richard Stones & Neil Matthew
(1996 Wrox Press ISBN 1-874416-68-0)

    This excellent book gave me the lowdown on many of the techniques
    used in the code.

Linux Man Page HOWTO by Jens Schweikhardt <schweikh@noc.dfn.de> and man(7)
by Rickard E. Faith <faith@cs.unc.edu> & David A. Wheeler <dwheeler@ida.org>
(http://www.linuxdoc.org/HOWTO/mini/Man-Page.html)

    These gave me the Linux standard by which to write my man page, plus
    many useful tips on producing documentation.

GNU Coding Standards by Richard Stallman 
(http://www.fsf.org/prep/standards_toc.html)

    This very thorough document provided some excellent pointers on how
    to write the code. I can't say I agree with the whole document,
    however, so the code in signature isn't entirely compliant.

GNU autoconf & automake manuals (http://www.fsf.org/manual/manual.html)

    These explained how to create a decent build environment.

Software Release Practice HOWTO v2.0 by Eric S. Raymond <esr@thyrsus.com>
(http://www.linuxdoc.org/HOWTO/Software-Release-Practice-HOWTO.html)

    This explained how to go about releasing the package.

Read INSTALL for details on how to compile and install the program and
its man page.

The software can be obtained from:

    http://www.caliban.org/files/signature/signature-0.14.tar.gz

RPM format packages can be obtained from:

    http://www.caliban.org/files/signature/signature-0.14-1.i386.rpm
    http://www.caliban.org/files/signature/signature-0.14-1.src.rpm

The source code is available under the GPL 2.0 licence.

-- 
Ian Macdonald <ian@caliban.org>


NEWS
Current Version: 0.14
See the file ChangeLog for the details of all changes.

User-visible changes from 0.13 to 0.14

* fixed bug that caused '<' placeholder to remain behind when there were
  backspaces or underscores in the text

User-visible changes from 0.12 to 0.13

* white space at the ends of lines is now trimmed
* ^H (backspace) and _ characters are now ignored in fortune text
* added a couple of debugging messages (we now report PID and FIFO name)
* fixed bug that caused FIFO and quote file not to be found when using -f and -q

User-visible changes from 0.11 to 0.12

* signature no longer exits when used with Pine. It still doesn't work,
  but it no longer exits either.

User-visible changes from 0.10 to 0.11

* external programs can now be called with arguments (shell quoting required)
* quote files can now use either ~ or % as a separator

User-visible changes from 0.04 to 0.10

* no-daemon mode is now selected with -n
* -d now provides debugging output on stderr
* support for quote files with -q
* bug fix for some fatal errors not clearing up temporary FIFO
* man page updated for quote files

User-visible changes from 0.03 to 0.04

* configure now checks for availability of offensive fortunes
  (Red Hat doesn't use them by default)
* man page updated to reflect the above

User-visible changes from 0.02 to 0.03

* when supplying an alternative program to fortune with the -p switch,
  the full path no longer has to be provided
* man page updated to reflect the above change
* usage message omitted -f in option list
* signature now forks into background on start-up. The -d switch has
  been added to prevent this for debugging purposes.

User-visible changes from 0.01 to 0.02

* default sig format string changed for kernel and architecture strings
  of different lengths
* zombie process build-up bug solved
* usage message tidied up


quote.c
/* quote.c -- quote file routines for signature

   $Id: quote.c,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#define _XOPEN_SOURCE

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "quote.h"

extern unsigned short int debug;

/* get a quote from an external quote file */
FILE *index_quote_file(char *quote_file, int *total_quotes,
		       struct sig_node **first_node)
{
    struct sig_node *node, *prev_node;
    unsigned short int sep_flag = 0, lf_flag = 0;
    FILE *quotes = NULL;

    /* initialise quote count */
    *total_quotes = 0;

    if (debug) {
	fprintf(stderr, "Opening quote file %s...\n", quote_file);
    }

    /* open quote file */
    if ((quotes = fopen(quote_file, "r")) == NULL) {
	perror(quote_file);
	unlink_fifo(EXIT_FAILURE);
    }
    /* create a linked list of quote offsets & lengths */
    node = (struct sig_node *) malloc(sizeof(struct sig_node));
    node->offset = 0;
    *first_node = node;
    while (!feof(quotes)) {
	switch (fgetc(quotes)) {
	case '~':
	case '%':
	    if (lf_flag) {
		/* set flag if this ~ follows a LF */
		sep_flag = 1;
		lf_flag = 0;
	    }
	    break;
	case '\n':
	    if (sep_flag) {
		/* we have a LF, preceded by a ~, preceded by a LF */
		node->length = ftell(quotes) - node->offset - 2;
		if (debug) {
		    fprintf(stderr,
			    "Quote # %5d:\toffset: %8d\tlength: %8d\n",
			    *total_quotes, node->offset, node->length);
		}
		/* make this node the previous one */
		prev_node = node;
		/* create a new node */
		node = (struct sig_node *) malloc(sizeof(struct sig_node));
		node->offset = ftell(quotes);
		/* point previous node to this one */
		prev_node->next_node = node;
		/* increment quote count */
		++(*total_quotes);
		sep_flag = 0;
	    }
	    lf_flag = 1;
	    break;
	default:
	    sep_flag = 0;
	    lf_flag = 0;
	}
    }
    return (quotes);
}

/* print a quote from the quote file */
int
print_quote(FILE * quotes, int total_quotes, struct sig_node *first_node)
{
    struct sig_node *node;
    char *quote_buf;
    int rand_sig;

    node = first_node;
    /* seed the random number generator on this PID -- this should be random
       enough for our purposes */
    srand(getpid());
    /* get a random number in the range 0 - total_quotes -- got this example
       from the NOTES section of rand(3) */
    rand_sig = (int) ((float) total_quotes * rand() / (RAND_MAX + 1.0));
    /* skip to the correct index */
    for (; rand_sig > 0; --rand_sig) {
	node = node->next_node;
    }
    /* seek to start of quote in file */
    fseek(quotes, node->offset, SEEK_SET);
    /* create and zero a buffer to hold the quote) */
    quote_buf = (char *) malloc(node->length + 1);
    memset(quote_buf, 0, node->length + 1);
    /* read the quote into the buffer */
    read(fileno(quotes), quote_buf, node->length);
    /* print quote */
    puts(quote_buf);
    /* free buffer */
    free(quote_buf);
    return (EXIT_SUCCESS);
}


quote.h
/* quote.h -- header file for quote.c

   $Id: quote.h,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


/* get a quote from an external quote file */
FILE *index_quote_file(char *quote_file, int *total_quotes,
		       struct sig_node **first_node);

/* print a quote from the quote file */
int print_quote(FILE * quotes, int total_quotes,
		struct sig_node *first_node);

/* remove the FIFO and exit */
extern void unlink_fifo(int status);


format.c
/* format.c -- formatting routine for signature

   $Id: format.c,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#define _XOPEN_SOURCE

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/utsname.h>

#include "format.h"

/* format signature for printing */
void format_sig(char *sig, char *buffer, int buf_size)
{
    struct utsname my_uname;
    char tmp_sig[BUF_SIZE];
    char *sig_ptr, *buf_ptr, *tmp_buf_ptr;
    int space_left, word_len, lf_flag;

    /* bollocks, this part is much easier in Perl! */
    lf_flag = 0;
    buf_ptr = buffer;
    /* terminate signature */
    *(sig + strlen(sig) - 1) = '\0';
    for (sig_ptr = sig; sig_ptr < sig + strlen(sig); ++sig_ptr) {
	if (buf_ptr >= buffer + buf_size) {
	    /* all of fortune printed, so shunt trailing white space away */
	    if (*sig_ptr == '<') {
		strcpy(sig_ptr, sig_ptr + 1);
		--sig_ptr;
	    }
	    continue;
	} else if (*sig_ptr != '<') {
	    /* in set text area, so fast forward to placeholder */
	    lf_flag = 1;
	    continue;
	} else if (*buf_ptr == '\b' || *buf_ptr == '_') {
	    ++buf_ptr;
	    --sig_ptr;
	    continue;
	} else if (*buf_ptr == '\n' || *buf_ptr == '\t' || *buf_ptr == ' ') {
	    if (!lf_flag) {
		/* replace a tab, space or LF with a space */
		*sig_ptr = ' ';
	    } else {
		/* on first placeholder char of line, so backtrack and don't
		   print anything */
		--sig_ptr;
	    }
	    ++buf_ptr;
	    continue;
	}

	lf_flag = 0;
	/* look ahead and calculate length of next word of fortune */
	tmp_buf_ptr = buf_ptr;
	while (*tmp_buf_ptr != ' ' && *tmp_buf_ptr != '\t'
	       && *tmp_buf_ptr != '\n' && *tmp_buf_ptr != '\b'
	       && *tmp_buf_ptr != '_') {
	    ++tmp_buf_ptr;
	}
	word_len = tmp_buf_ptr - buf_ptr;

	/* look ahead and calculate how much space we have on the current
	   line for the next word of fortune */
	tmp_buf_ptr = sig_ptr;
	space_left = 0;
	while (*tmp_buf_ptr++ == '<') {
	    if (++space_left == word_len) {
		break;
	    }
	}

	/* if next word won't fit on current line, shunt trailing white space 
	   away */
	if (space_left < word_len) {
	    strcpy(sig_ptr - 1, sig_ptr + space_left);
	    continue;
	}

	/* copy word from fortune to signature */
	strncpy(sig_ptr, buf_ptr, word_len);
	sig_ptr += word_len - 1;
	buf_ptr += word_len;
    }

    /* get OS information */
    if (uname(&my_uname) == -1) {
	perror("uname() failed");
	unlink_fifo(EXIT_FAILURE);
    }

    /* make a copy of the sig for use as a format string */
    strcpy(tmp_sig, sig);
    /* add OS information to signature */
    sprintf(sig, tmp_sig, my_uname.release, my_uname.machine);
}


format.h

/* format.h -- header file for format.c

   $Id: format.h,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


/* format signature for printing */
void format_sig(char *sig, char *buffer, int buf_size);

/* remove the FIFO and exit */
extern void unlink_fifo(int status);


signature.h
/* signature.h -- header file for signature.c

   $Id: signature.h,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


/* display version information */
void version(void);

/* display help message */
void help(int status);

/* remove the FIFO and exit */
void unlink_fifo(int status);

/* when we catch a signal */
void caught_signal(int sig);

/* ignore SIGPIPE */
void ignore_sigpipe(int sig);

/* set up FIFO */
void init_fifo(unsigned short int remove_flag);

/* install signal handler */
void install_sig_handler(void);

/* format signature for printing */
extern void format_sig(char *sig, char *buffer, int buf_size);

/* get a quote from an external quote file */
extern FILE *index_quote_file(char *quote_file, int *total_quotes,
			      struct sig_node **first_node);

/* print a quote from the quote file */
extern int print_quote(FILE * quotes, int total_quotes,
		       struct sig_node *first_node);


signature.c
/* signature.c -- create a dynamic signature facility.

   $Id: signature.c,v 1.1.1.1 2009/10/24 17:48:09 reynolds Exp $

   Copyright (C) Ian Macdonald <ian@caliban.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#define _XOPEN_SOURCE

#include <config.h>

#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "signature.h"

/* name of default signature file */
#define DEFAULT_FIFO ".signature"

/* *INDENT-OFF* */

/* text of default signature */
#define SIGNATURE "\
Linux, the choice          | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n\
of a GNU generation   -o)  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n\
Kernel %-9.9s       /\\  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n\
on a %-6.6s           _\\_v | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n\
                           | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n"

/* *INDENT-ON* */

static char fifo[BUF_SIZE];
static char quote_file[BUF_SIZE];
unsigned short int debug = 0;

/* display version information */
void version(void)
{
    printf("%s version %s ("
	   __DATE__ " " __TIME__ ")\n\n"
	   "Copyright (C) 1999-2000 Ian Macdonald <ian@caliban.org>\n"
	   "This is free software; see the source for copying conditions.\n"
	   "There is NO warranty; not even for MERCHANTABILITY or FITNESS\n"
	   "FOR A PARTICULAR PURPOSE, to the extent permitted by law.\n",
	   PACKAGE, VERSION);
    exit(EXIT_SUCCESS);
}

/* display help message */
void help(int status)
{
    printf
	("Usage: %s [ -d ] [ -f FIFO ] [ -n ] [ -p program | -q quote file ]\n"
	 "                 [ -r ] [ -t template ]\n"
	 "       %s -h | -v\n\n"
	 "-d              print debugging messages on stderr\n"
	 "-f <FIFO>       use this file as FIFO instead of ~/.signature\n"
	 "-h              display this help message\n"
	 "-n              don't fork into background as daemon\n"
	 "-p <program>    use this program instead of fortune(6)\n"
	 "-q <quote file> pluck a random quote from this file instead of from fortune(6)\n"
	 "-r              force removal of old signature FIFO if present\n"
	 "-t <template>   use this signature template instead of default\n"
	 "-v              display version information\n\n"
	 "E-mail bug reports and diffs to: ian@caliban.org\n", PACKAGE,
	 PACKAGE);
    exit(status);
}

/* remove the FIFO and exit */
void unlink_fifo(int status)
{
    if (debug) {
	fprintf(stderr, "Unlinking FIFO...\n");
    }
    unlink(fifo);
    exit(status);
}

/* when we catch a signal */
void caught_signal(int sig)
{
    if (debug) {
	fprintf(stderr, "Caught signal %d. Clearing up...\n", sig);
    }
    unlink_fifo(EXIT_SUCCESS);
}

/* ignore SIGPIPE */
void ignore_sigpipe(int sig)
{
    if (debug) {
	fprintf(stderr, "Caught a SIGPIPE. Using Pine by any chance?\n");
    }
}

/* set up FIFO */
void init_fifo(unsigned short int remove_flag)
{
    if (access(fifo, F_OK) == 0 && remove_flag) {
	/* FIFO already exists */
	if (unlink(fifo) == -1) {
	    fprintf(stderr, "Couldn't unlink %s\n", fifo);
	    exit(EXIT_FAILURE);
	}
    }
    if (mkfifo(fifo, 0644) == -1) {
	fprintf(stderr, "Couldn't create %s as FIFO\n", fifo);
	exit(EXIT_FAILURE);
    }
}

/* install signal handler */
void install_sig_handler(void)
{
    struct sigaction act;

    /* install signal handler */
    act.sa_handler = caught_signal;
    sigemptyset(&act.sa_mask);
    act.sa_flags = 0;
    sigaction(SIGINT, &act, 0);
    sigaction(SIGHUP, &act, 0);
    sigaction(SIGTERM, &act, 0);
    act.sa_handler = ignore_sigpipe;
    sigemptyset(&act.sa_mask);
    sigaction(SIGPIPE, &act, 0);
}

int main(int argc, char *argv[])
{
    char signature[BUF_SIZE], sig_buf[BUF_SIZE], fortune[BUF_SIZE];
    char *producer = NULL, *prodarg = NULL;
    char *prodargs[MAX_PROG_ARGS];
    char *slash_ptr, *cwd_ptr;
    int prodargc = 0;
    unsigned short int no_fork = 0;
    int exit_status = 0;
    int sig_fd, sig_buf_len = 0;
    int opt, pid;
    int file_pipes[2], fifo_fd, num_read, remove_flag = 0;
    int total = 0;
    FILE *quote_stream;
    struct sig_node *first_node;
    /* got to use a pointer to a pointer to a struct here. Don't like this -- 
       too complicated */
    struct sig_node **first_node_ptr = &first_node;

    memset(&fifo, 0, BUF_SIZE);
    memset(&quote_file, 0, BUF_SIZE);
    /* get option switches */
    while ((opt = getopt(argc, argv, "df:hnp:q:rt:v")) != -1) {
	switch (opt) {
	case 'd':
	    /* set debugging flag */
	    debug = 1;
	    break;
	case 'f':
	    /* get location of FIFO */
	    slash_ptr = strrchr(optarg, '/');
	    if (slash_ptr == NULL || *optarg == '.') {
		/* make path absolute */
		cwd_ptr = getcwd(NULL, BUF_SIZE);
		sprintf(fifo, "%s/%s", cwd_ptr, optarg);
		free(cwd_ptr);
	    } else {
		strcpy(fifo, optarg);
	    }
	    if (debug) {
		fprintf(stderr, "FIFO is %s\n", fifo);
	    }
	    break;
	case 'h':
	    help(EXIT_SUCCESS);
	    break;
	case 'n':
	    no_fork = 1;
	    break;
	case 'p':
	    if (quote_file[0] != '\0') {
		help(EXIT_FAILURE);
	    }
	    producer = optarg;
	    prodargs[prodargc++] = strtok(producer, " ");
	    while ((prodarg = strtok(NULL, " ")) != NULL) {
		prodargs[prodargc++] = prodarg;
		if (prodargc > MAX_PROG_ARGS) {
		    fprintf(stderr, "More than %d arguments to %s. "
			    "Excess arguments ignored...\n",
			    MAX_PROG_ARGS, producer);
		    break;
		}
	    }
	    prodargs[prodargc] = NULL;
	    /* check for existence of program by forking and then trying to
	       exec() it in the child */
	    pid = fork();
	    switch (pid) {
	    case -1:		/* oh well */
		perror("Couldn't fork() a child process");
		exit(EXIT_FAILURE);
	    case 0:		/* in child */
		/* close stdout */
		close(1);
		execlp(producer, producer, (char *) 0);
		exit(EXIT_FAILURE);
	    default:
		waitpid(pid, &exit_status, 0);
		if (exit_status != EXIT_SUCCESS) {
		    fprintf(stderr, "Program %s doesn't seem to exist\n",
			    producer);
		    exit(EXIT_FAILURE);
		}
	    }
	    break;
	case 'q':
	    if (producer) {
		help(EXIT_FAILURE);
	    }
	    slash_ptr = strrchr(optarg, '/');
	    if (slash_ptr == NULL || *optarg == '.') {
		/* make path absolute */
		/* get the current directory */
		cwd_ptr = getcwd(NULL, BUF_SIZE);
		/* get absolute path of quote_file */
		sprintf(quote_file, "%s/%s", cwd_ptr, optarg);
		free(cwd_ptr);
	    } else {
		/* get a quote from an external file instead of fortune */
		strcpy(quote_file, optarg);
	    }
	    break;
	case 'r':
	    remove_flag = 1;
	    break;
	case 't':
	    if (debug) {
		fprintf(stderr, "Opening template file %s...\n", optarg);
	    }
	    sig_fd = open(optarg, O_RDONLY);
	    if (sig_fd == -1) {
		fprintf(stderr, "Couldn't open template %s for reading\n",
			optarg);
		exit(EXIT_FAILURE);
	    }
	    memset(&sig_buf, 0, BUF_SIZE);
	    sig_buf_len = read(sig_fd, &sig_buf, BUF_SIZE);
	    if (sig_buf_len == -1) {
		perror("Failed to read() template file");
		exit(EXIT_FAILURE);
	    }
	    close(sig_fd);
	    break;
	case 'v':
	    version();
	    break;
	case ':':
	    fprintf(stderr, "Option needs a value\n");
	    break;
	case '?':
	    exit(EXIT_FAILURE);
	}
    }
    if (fifo[0] == '\0') {
	/* set up the path to default FIFO */
	sprintf(fifo, "%s/%s", getenv("HOME"), DEFAULT_FIFO);
    }
    /* initialise FIFO */
    init_fifo(remove_flag);
    /* install signal handler */
    install_sig_handler();

    if (!no_fork) {
	/* fork off into background */
	pid = fork();
	switch (pid) {
	case -1:		/* oh well */
	    perror("Couldn't fork() a child process");
	    unlink_fifo(EXIT_FAILURE);
	case 0:		/* in child */
	    /* become a process group and session group leader */
	    setsid();
	    /* fork off into background */
	    pid = fork();
	    switch (pid) {
	    case -1:		/* oh well */
		perror("Couldn't fork a child process");
		unlink_fifo(EXIT_FAILURE);
	    case 0:		/* in child */
		if (debug) {
		    fprintf(stderr, "PID is %d\n", getpid());
		}
		/* ensure our process doesn't keep any directory in use */
		chdir("/");
		break;
	    default:
		exit(EXIT_SUCCESS);
	    }
	    break;
	default:
	    exit(EXIT_SUCCESS);
	}
    }

    if (quote_file[0] != '\0') {
	quote_stream =
	    index_quote_file(quote_file, &total, first_node_ptr);
    }

    /* start main loop */
    while (1) {
	memset(&signature, 0, BUF_SIZE);
	if (sig_buf_len > 0) {
	    /* use user-supplied template */
	    strcpy(signature, sig_buf);
	} else {
	    /* use stock signature */
	    strcpy(signature, SIGNATURE);
	}

	if (access(fifo, F_OK) != 0) {
	    fprintf(stderr, "%s seems to have disappeared\n", fifo);
	    exit(EXIT_FAILURE);
	}

	/* open FIFO for writing */
	fifo_fd = open(fifo, O_WRONLY);
	if (fifo_fd == -1) {
	    fprintf(stderr, "Couldn't open %s for writing\n", fifo);
	    unlink_fifo(EXIT_FAILURE);
	}
	/* create a pipe */
	if (pipe(file_pipes) != 0) {
	    perror("Couldn't create pipe() for passing signature");
	    unlink_fifo(EXIT_FAILURE);
	}
	/* fork off a child */
	pid = fork();
	switch (pid) {
	case -1:		/* oh well */
	    perror("Couldn't fork() a child process");
	    unlink_fifo(EXIT_FAILURE);
	case 0:		/* in child */
	    /* close stdout */
	    close(1);
	    /* duplicate write fd as stdout */
	    dup(file_pipes[1]);
	    /* close write fd */
	    close(file_pipes[1]);
	    /* close read fd */
	    close(file_pipes[0]);

	    /* check if we need to print from a quote file */
	    if (quote_file[0] != '\0') {
		exit(print_quote(quote_stream, total, first_node));
	    } else if (!producer) {	/* call fortune */
		if (debug) {
		    fprintf(stderr, "Calling %s...\n", FORTUNE);
		}
		/* send fortune cookie to write fh */
		execl(FORTUNE, "fortune", "-s"
#ifdef OFFENSIVE_FORTUNES
		      "a"
#endif
		      , (char *) 0);
		/* we only reach this point if exec() didn't work */
		fprintf(stderr, "Couldn't exec %s\n", FORTUNE);
		exit(EXIT_FAILURE);
	    } else {		/* external program */
		if (debug) {
		    fprintf(stderr, "Calling %s...\n", producer);
		}
		/* send producer's output to write fh */
		execvp(producer, prodargs);
		/* we only reach this point if exec() didn't work */
		fprintf(stderr, "Couldn't exec %s\n", producer);
		exit(EXIT_FAILURE);
	    }
	default:		/* in parent */
	    if (debug) {
		fprintf(stderr, "Child process PID %d...\n", pid);
	    }
	    /* clean up after child */
	    waitpid(pid, (int *) 0, 0);
	    /* close write fd */
	    close(file_pipes[1]);
	    /* read fortune cookie through pipe */
	    num_read = read(file_pipes[0], fortune, PIPE_BUF);
	    if (!num_read) {
		/* exec() didn't work: quit parent */
		perror("Error in read() of FIFO");
		unlink_fifo(EXIT_FAILURE);
	    }
	    /* close read fd */
	    close(file_pipes[0]);
	    /* do all the hard stuff */
	    format_sig(signature, fortune, num_read);
	    /* send it to the FIFO */
	    write(fifo_fd, signature, strlen(signature));
	    /* close FIFO */
	    close(fifo_fd);
	}
	/* loop for next signature */
	sleep(1);
    }
}
